src/GafferScene/IECoreScenePreview/MeshAlgoTessellate.cpp

```diff
@@ -0,0 +1,1491 @@
+	// them to those shared points computed above for the boundary.
+	// This is easy, because interior points are never shared.
+	if( numInteriorCoords )
+	{
+		for (int i = 0; i < numInteriorCoords; ++i)
+		{
+			evaluateSurface<T>(
+				surface, buffers.patchPoints.data(), (float*)&tessInteriorCoords[ i ],
+				outOffset++, out, outNormals
+			);
+		}
+	}
+
+	if( setup.m_outIndicesWritable )
+	{
+		// Write out the vertex indices for the faces coming from all our tessellated facets.
+
+		// If we are writing out quad facets, but the face is irregular, and the tessellation rate is odd,
+		// then OpenSubDiv will write out some quad facets that are actually triangles, labelled with one vert
+		// set to -1. In order to output accurate topology, we need to collapse this list, removing -1s, and
+		// adjusting the vertex counts of corresponding faces.
+		const bool needsCollapse = tessPattern.GetFacetSize() == 4 && fVerts.size() != 4 && ( tessUniformRate & 1 );
+
+		int *outIndices;
+		if( needsCollapse )
+		{
+			// If we need to collapse vertex ids, then we have to put the vertex ids in a temporary buffer
+			// instead of writing straight to the final buffer.
+			buffers.collapseIndices.resize( tessPattern.GetNumFacets() * 4 );
+			outIndices = buffers.collapseIndices.data();
+		}
+		else
+		{
+			outIndices = &setup.m_outIndicesWritable[outIndicesIndex];
+		}
+
+
+		int tessInteriorOffset = outOffset - numOutCoords;
+		tessPattern.GetFacets( outIndices );
+
+
+		// GetFacets generates coordinate indices used by the facets are local
+		// to the face (i.e. they range from [0..N-1], where N is the
+		// number of coordinates in the pattern) and so need to be offset
+		// to refer to global vertex indices.
+		//
+		// Whereas the edge gets overwritten with the boundary indices we've
+		// computed, which are already global. TransformFacetCoordIndices does
+		// both.
+
+		// TransformFacetCoordIndices seems to incorrectly not be labelled as const, so we cheat
+		// with a const_cast
+		const_cast< OSDB::Tessellation* >( &tessPattern )->TransformFacetCoordIndices(
+			outIndices, buffers.boundaryIndices.data(), tessInteriorOffset
+		);
+
+		if( needsCollapse )
+		{
+			// If some of our quads are actually tris, we need to scan through looking for indices of -1,
+			// and not copying those to the actual output. Whenever we find a -1, we decrease the
+			// verticesPerFace count for the corresponding face, so that we output actual triangles,
+			// instead of degenerate quads.
+			for( unsigned int i = 0; i < buffers.collapseIndices.size(); i++ )
+			{
+				if( buffers.collapseIndices[i] != -1 )
+				{
+					setup.m_outIndicesWritable[outIndicesIndex++] = buffers.collapseIndices[i];
+				}
+				else
+				{
+					// Update the verticesPerFace count to handle that we've removed a vert, and this face is
+					// no longer a quad.
+					if( outVerticesPerFace )
+					{
+						outVerticesPerFace[ i / 4 ]--;
+					}
+				}
+			}
```

Comment:
There's some coverage here:

https://github.com/GafferHQ/gaffer/blob/c40022fe390715e7d920df83436c21da3a4f2ab8/python/GafferSceneTest/MeshTessellateTest.py#L367

And I snuck some coverage into testNonManifold, which features irregular faces and a tessellation rate of 3.

But it sounds like you're thinking of something that would fail in the simplest possible way if this feature was missing - I can add a simple test case based on createTestData, which won't test as much as testNonManifold, but would be an obvious failure.

Reply:
Yeah, I think there's a lot of merit in small tests that isolate one thing - they're easier to debug than huge functions that test everything and then fail somewhere in the middle...